File: NaSt2D-2.0/Tools/f/neighbor.f
c*********************************************************
c     Program to search the neighbor points of isolines
c     -1. -1. -1. labelled the input end
c*--------------------------------------------------------
c     Programm zur Suche der Nachbarpunkte von Isolinien
c     -1. -1. -1. kennzeichnet das Eingabeende
c*********************************************************
      program Neighbor
c
      parameter (n=10000)
      dimension a(n),b(n),c(n),d(n),e(n)
      real min
      open (unit=1,file='ein')
      open (unit=2,file='aus')
c
      do 10 i=1,n
      read(1,*) a(i),b(i),c(i) 
      d(i)=0.
      e(i)=1000.
      if(a(i).eq.-1..and.b(i).eq.-1..and.c(i).eq.-1.) goto 20
   10 continue
   20 in=i-1
      write(2,*) a(1),b(1),c(1)

      iprint=1
      do 70 ii=1,in-1
      min=e(1)
      do 30 i=2,in
   30 e(i)=sqrt((a(i)-a(iprint))**2 + (b(i)-b(iprint))**2)
      do 40 i=2,in
   40 if(d(i).eq.0..and.e(i).lt.min) min=e(i)
      do 50 i=2,in
      if(d(i).eq.0..and.e(i).eq.min) goto 60
   50 continue
   60 write(2,*) a(i),b(i),c(i)
      iprint=i
      d(i)=1.
   70 continue
 
      stop 'ok'
      end
c**********************************************************

File: NaSt2D-2.0/Tools/c/neighbor.c
/*     neighbor.c */
/********************************************************************/
/*     Program to search the neighbor points of isolines            */
/*     -1. -1. -1. labelled the input end                           */
/*------------------------------------------------------------------*/
/*     Programm zur Suche der Nachbarpunkte von Isolinien           */
/*     -1. -1. -1. kennzeichnet das Eingabeende                     */
/********************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define VOL1 1000   /* 1st dimension row    - Zeilenindex  i */
#define VOL2 5      /* 2nd dimension column - Spaltenindex j */

int main(void) 
{
   int    i=0,in=0,j,ii,iprint;
   float  array[VOL1][VOL2],min;
   FILE   *fpein;
   FILE   *fpaus;
   char   *input="ein";                            /* input file  */
   char   *output="aus";                           /* output file */

   fpein=fopen(input,"r");                         /* opening the files */
   fpaus=fopen(output,"w");
   if ((fpein==NULL)||(fpaus==NULL)) {
      printf("Error on opening the files\n");
      return(-1);                                  /* EXIT_ERROR_input_file */
                                     }

   for(i=0; i < VOL1; i++) {                       /* read in array */
      for(j=0; j < VOL2-2; j++) fscanf(fpein,"%f ", &array[i][j]);
      array[i][3]=0.;
      array[i][4]=1000.;
      in=i-1;
      if((array[i][0] == array[i][1]) && (array[i][0] == array[i][2])) break;
                           }

   i=0;                                            /* Start point */
   array[i][3]=1.;
   for(j=0; j < VOL2-2; j++) fprintf(fpaus,"%3.8f ",array[i][j]);
   fprintf(fpaus,"\n");

   iprint=0;
   for(ii=1; ii<=in+1; ii++) {
   min=array[0][4];
   for(i=1; i<=in; i++) {                          /* calculation of distances */
   array[i][4]=sqrt((array[i][0]-array[iprint][0])*(array[i][0]-array[iprint][0]) + 
                    (array[i][1]-array[iprint][1])*(array[i][1]-array[iprint][1]));
   if(array[i][3] == 0. && array[i][4] < min) min=array[i][4];
                        }

   for(i=1; i<=in; i++) {                          /* write at minimum distanc */
   if(array[i][3] == 0. && array[i][4] == min) {
      iprint=i;
      for(j=0; j < VOL2-2; j++) fprintf(fpaus,"%3.8f ",array[i][j]);	
      fprintf(fpaus,"\n");
      array[i][3]=1.;
                                               }
                        }
                             }

   printf("neighbor EXIT_SUCCESS\n");
   fclose(fpein);
   fclose(fpaus);
   return EXIT_SUCCESS;
}














